### PR TITLE
feat: Data source cache

### PR DESCRIPTION
**Pull request by scsmithr**
Adds caching of external database table info.

```
> CREATE EXTERNAL DATABASE my_pg
::: FROM postgres OPTIONS (
:::   host = 'pg.demo.glaredb.com',
:::   port = '5432',
:::   user = 'demo',
:::   password = 'demo',
:::   database = 'postgres',
::: );
Database created
> select * from cache_external_database_tables();
┌────────────────────────────────┬─────────────────────────┬─────────────────────────┐
│ system_operation_name          │ started_at              │ finished_at             │
│ ──                             │ ──                      │ ──                      │
│ Utf8                           │ Date64                  │ Date64                  │
╞════════════════════════════════╪═════════════════════════╪═════════════════════════╡
│ cache_external_database_tables │ 2023-12-05T03:10:17.086 │ 2023-12-05T03:10:42.756 │
└────────────────────────────────┴─────────────────────────┴─────────────────────────┘
> select * from glare_catalog.cached_external_database_tables;
┌──────────────┬─────────────┬──────────────────┬───────────────────────┬────────────────────────────────────┐
│ database_oid │ schema_name │ table_name       │ column_name           │ data_type                          │
│           ── │ ──          │ ──               │ ──                    │ ──                                 │
│        Int32 │ Utf8        │ Utf8             │ Utf8                  │ Utf8                               │
╞══════════════╪═════════════╪══════════════════╪═══════════════════════╪════════════════════════════════════╡
│        20000 │ pg_catalog  │ pg_stat_archiver │ archived_count        │ Int64                              │
│        20000 │ pg_catalog  │ pg_stat_archiver │ last_archived_wal     │ Utf8                               │
│        20000 │ pg_catalog  │ pg_stat_archiver │ last_archived_time    │ Timestamp(Nanosecond, Some("UTC")) │
│        20000 │ pg_catalog  │ pg_stat_archiver │ failed_count          │ Int64                              │
│        20000 │ pg_catalog  │ pg_stat_archiver │ last_failed_wal       │ Utf8                               │
│        20000 │ pg_catalog  │ pg_stat_archiver │ last_failed_time      │ Timestamp(Nanosecond, Some("UTC")) │
│        20000 │ pg_catalog  │ pg_stat_archiver │ stats_reset           │ Timestamp(Nanosecond, Some("UTC")) │
│        20000 │ pg_catalog  │ pg_stat_bgwriter │ checkpoints_timed     │ Int64                              │
│        20000 │ pg_catalog  │ pg_stat_bgwriter │ checkpoints_req       │ Int64                              │
│        20000 │ pg_catalog  │ pg_stat_bgwriter │ checkpoint_write_time │ Float64                            │
│            … │ …           │ …                │ …                     │ …                                  │
│        20000 │ public      │ orders           │ o_comment             │ Utf8                               │
│        20000 │ public      │ part             │ p_partkey             │ Int32                              │
│        20000 │ public      │ part             │ p_name                │ Utf8                               │
│        20000 │ public      │ part             │ p_mfgr                │ Utf8                               │
│        20000 │ public      │ part             │ p_brand               │ Utf8                               │
│        20000 │ public      │ part             │ p_type                │ Utf8                               │
│        20000 │ public      │ part             │ p_size                │ Int32                              │
│        20000 │ public      │ part             │ p_container           │ Utf8                               │
│        20000 │ public      │ part             │ p_retailprice         │ Decimal128(38, 9)                  │
│        20000 │ public      │ part             │ p_comment             │ Utf8                               │
└──────────────┴─────────────┴──────────────────┴───────────────────────┴────────────────────────────────────┘
 131 rows (20 shown)
```

### CODE REVIEW

File: crates/sqlbuiltins/src/builtins.rs [hunk 1]
```diff
@@ -34,11 +34,33 @@ pub const POSTGRES_SCHEMA: &str = "pg_catalog";
...
+/// depends on a schema.
+///
+/// Builtin tables have stable OIDs since some depend on data written to disk.
+///
+/// First glaredb builtin OID: 16384
+/// First user object OID: 20000
+///
+/// This means we have ~3600 OIDs to play with for builtin objects. Note that
+/// once a builtin object is given a stable OID, it **must not** be changed ever
+/// (unless you're the person willing to write a migration system).
```
Comment (tychoish) [inline crates/sqlbuiltins/src/builtins.rs:49]: which we should do in a while?

File: crates/sqlbuiltins/src/builtins.rs [hunk 2]
```diff
@@ -34,11 +34,33 @@ pub const POSTGRES_SCHEMA: &str = "pg_catalog";
...
-pub const FIRST_NON_SCHEMA_ID: u32 = FIRST_GLAREDB_BUILTIN_ID + 100;
+/// Builtin schemas have stable OIDs since everything (builtin and user objects)
+/// depends on a schema.
+///
+/// Builtin tables have stable OIDs since some depend on data written to disk.
+///
+/// First glaredb builtin OID: 16384
+/// First user object OID: 20000
+///
+/// This means we have ~3600 OIDs to play with for builtin objects. Note that
```
Comment (tychoish) [inline crates/sqlbuiltins/src/builtins.rs:47]: why do they have to be sequentially grouped?

File: crates/sqlbuiltins/src/builtins.rs (hunk 1 above)
Comment (scsmithr) [reply crates/sqlbuiltins/src/builtins.rs:49]: Likely, just want to try to delay it as much as possible to better understand what needs to happen for a system like that.

File: crates/sqlbuiltins/src/builtins.rs (hunk 2 above)
Comment (scsmithr) [reply crates/sqlbuiltins/src/builtins.rs:47]: They don't, it's just easier to track what the next OID should be when adding a table/schema/whatever. If we have more than 16 builtin schemas (unlikely), we can just jump forward to some other available OID.

File: crates/sqlbuiltins/src/builtins.rs (hunk 1 above)
Comment (tychoish) [reply crates/sqlbuiltins/src/builtins.rs:49]: 1000% agree.

**Comment by tychoish** [comment]
What's the relationship of this pr to #2201

File: crates/datafusion_ext/src/system.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+use datafusion::execution::TaskContext;
+use futures::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+
+/// A system operation can execute an arbitrary operation.
+///
+/// This should be focused on operations that do not require user interactions
+/// (e.g. background operations like optimizing tables or running caching jobs).
+pub trait SystemOperation: Sync + Send {
```
Comment (universalmind303) [inline crates/datafusion_ext/src/system.rs:11]: Would we ever have an unknown number of `SystemOperation`'s, or custom operations? Considering we are dealing with an exhaustive list of operations, I'm wondering if it'd make sense to use an enum here instead?

File: crates/sqlbuiltins/src/functions/table/system/cache_external_tables.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+use crate::builtins::GLARE_CACHED_EXTERNAL_DATABASE_TABLES;
+use crate::functions::table::virtual_listing::get_virtual_lister_for_external_db;
+
+#[derive(Debug, Clone, Copy)]
+pub struct CacheExternalDatabaseTables;
+
+impl ConstBuiltinFunction for CacheExternalDatabaseTables {
+    const NAME: &'static str = "cache_external_database_tables";
+    const DESCRIPTION: &'static str = "Cache tables from external databases.";
+    const EXAMPLE: &'static str = "select * from cache_external_database_tables();";
```
Comment (universalmind303) [inline crates/sqlbuiltins/src/functions/table/system/cache_external_tables.rs:40]: Something for another day, but it'd be cool to support the `CALL` syntax for this.


while valid, `select * from cache_external_database_tables();` seems weird to me as im selecting from an action.

`CALL` seems fitting for something that is just executing a function call. 

```sql
CALL cache_external_database_tables();`
```

File: crates/datafusion_ext/src/system.rs (hunk 3 above)
Comment (scsmithr) [reply crates/datafusion_ext/src/system.rs:11]: Fair, will change

File: crates/sqlbuiltins/src/functions/table/system/cache_external_tables.rs (hunk 4 above)
Comment (scsmithr) [reply crates/sqlbuiltins/src/functions/table/system/cache_external_tables.rs:40]: I agree, I originally wanted to use CALL, but sqlparser doesn't support it.

**Review by tychoish — APPROVED** [review]

**Comment by scsmithr** [comment]
Force merging, CI failure due to https://github.com/GlareDB/glaredb/issues/2168

**Pull request merged**